File: csharp/protein-translation/ProteinTranslation.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class ProteinTranslation
{
    private static readonly Dictionary<string, string> Proteins = new Dictionary<string, string>
    {
        ["AUG"] = "Methionine",
        ["UUU"] = "Phenylalanine",
        ["UUC"] = "Phenylalanine",
        ["UUA"] = "Leucine",
        ["UUG"] = "Leucine",
        ["UUG"] = "Leucine",
        ["UCU"] = "Serine",
        ["UCC"] = "Serine",
        ["UCA"] = "Serine",
        ["UCG"] = "Serine",
        ["UAU"] = "Tyrosine",
        ["UAC"] = "Tyrosine",
        ["UGU"] = "Cysteine",
        ["UGC"] = "Cysteine",
        ["UGG"] = "Tryptophan",
        ["UAA"] = "STOP",
        ["UAG"] = "STOP",
        ["UGA"] = "STOP",

    };

    public static string[] Translate(string sequence)
    {
        IEnumerable<char> Project(string source, int word)
        {
            int index = word * 3;
            yield return source.ElementAt(index);
            yield return source.ElementAt(index + 1);
            yield return source.ElementAt(index + 2);
        }

        return Enumerable.Range(0, sequence.Length / 3)
                .Select(i => new String(Project(sequence, i).ToArray()))
                .Select(i => Proteins.TryGetValue(i, out string protein) ? protein : throw new Exception($"Unknown codon {i} detected"))
                .TakeWhile(i => i != "STOP")
                .ToArray();


    }
}

File: python/protein-translation/protein_translation.py
"""exercise"""
STEP = 3
Codons = {"AUG":("Methionine",False),
          "UUC":("Phenylalanine",False),
          "UUU":("Phenylalanine",False),
          "UUA":( "Leucine",False),
          "UUG":( "Leucine",False),
          "UCU":( "Serine",False),
          "UCC":( "Serine",False),
          "UCA":( "Serine",False),
          "UCG":( "Serine",False),
          "UAU":( "Tyrosine",False),
          "UAC":( "Tyrosine",False),
          "UGU":( "Cysteine",False),
          "UGC":( "Cysteine",False),
          "UGG":( "Tryptophan",False),
          "UAA": ("",True),
          "UAG": ("",True),
          "UGA": ("",True)
          }

def proteins(strand):
    """
    :param strand:
    :return:
    """

    codons = [Codons[strand[index:index+STEP]] for index in range(0,len(strand),STEP)]
    acids = []
    done = False
    i = 0
    while not done:
        if i >= len(codons):
            done = True
        elif codons[i][1]:
            done = True
        else:
            acids.append(codons[i][0])
            i +=1
    return acids
